### ISSUE TITLE
Invalidate deduplication window after ack

### ISSUE THREAD

**Issue by ro0NL** [nats-io/nats-server#5142]
### Proposed change

We noticed messages are kept being deduplicated by Msg-ID for the full window time, even afer acking them.

I find this a bit hard to reason about, since it's practically a form of message lost.

Im wondering if invalidating from the client while acking is considerable?

### Use case

Real-time product sync, where the 2nd message after acking the 1st messsage can lead to different sync result

**Comment by ripienaar** [comment]
What kind of stream are you using WQ? Limits? For all but WQ many consumers can get the same message at different times etc and the fact that it was acked has no relevance to the stream.

For a WQ, the message will be removed after ack and then I would still think dedup should persist within the window specifically in a WQ because you want to avoid re-running already ran jobs.

**Comment by ro0NL** [comment]
WG yes.

> you want to avoid re-running already ran jobs

in this case no, the previous job was acked. the new job is a fresh dispatch

**Comment by ripienaar** [comment]
Why would you have 2 jobs with the same ID? This seems like a bug :) The purpose of dedup is to find duplicates using a unique sequence, if you're adding non unique sequences then that's the cause of the problem, use unique sequences.

**Comment by ro0NL** [comment]
they are real time updates, so we have 2 jobs because 2 users modified the same product

**Comment by ripienaar** [comment]
Then you need to have a job ID per job.  The purpose of this is to dedup on unique ids.  If you do not have unique IDs then this feature is not appropriate to use for your purpose.

**Comment by ripienaar** [comment]
> i agree 👍 i was hoping we could dedup like this at server level in the future. My usecase/reasoning is legit no? Though im not sure how it plays out technically.

Your use case is legit sure, but you are using the wrong tool to solve the problem if your IDs are not unique.  Changing the unique-id-requiring feature to not do that breaks the feature :)

> The alternative route for us is to have 0s dedup window, and then deduplicate in the worker.
> 
> Or dedup at client level, keeping an extra bookkeeping in redis or so.
> 
> Both would not be needed if we could invalidate the dedup window during ack 👼

Surely you can construct IDs thats a combination of product and user id? This would dedup the action on a per product+user dimension allowing multiple users to update one product and still protecting against those instructions being duplicated on a publish retry.

**Comment by ro0NL** [comment]
Im afraid our system is too generic to have fine grained message ID control for now. It's really simple as: "product changed event" => "dispatch product update".

> this feature exist to protect against publish retries landing in the stream multiple times

you mean a burst of retries done by the client? I agree those should be dedupped within the time window, my proposal would be to opt-in for invalidation during ack.

**Comment by ripienaar** [comment]
> Im afraid our system is too generic to have fine grained message ID control for now. It's really simple as: "product changed event" => "dispatch product update".

You definitely could just have a uuid as a event ID.

> > this feature exist to protect against publish retries landing in the stream multiple times
> 
> you mean a burst of retries done by the client? I agree those should be dedupped within the time window, my proposal would be to opt-in for invalidation during ack.

Yes, you publish to the stream but get an error because the ack was not received.  So you retry the publish, get the ack and all is good.

The question is why did you not get the ACK, it could be:

 * The server got your message, stored it, sent the ack and the ack got lost
 * The server did not get your messages so no ack was sent

Your only course of action is to retry the publish exactly as it was with the same message ID.  This feature exist to ensure that the first case - where the server got the message but the ack got lost - does not result in a duplicate message.

That's the purpose of it and for this purpose, even Acked messages need to be deduped.

**Comment by ro0NL** [comment]
> You definitely could just have a uuid as a event ID

i think so :')

**Comment by jnmoyne** [comment]
FWIW, and besides the whole very valid conversation about the validity of rejecting duplicates only for messages that haven't been ack'd yet you can get that exact behavior by using the ['infinite deduplication window'](https://nats.io/blog/new-per-subject-discard-policy/) technique on a working queue.

**Comment by jnmoyne** [comment]
Though normally you would not want to use the infinite deduplication on a working queue (for the very reasons @ripienaar mentions) and you would first record the messages in an 'audit' (not WQ or Interest) stream with infinite deduplication and consume from a working queue stream that would source from that audit stream.

**Comment by ro0NL** [comment]
ive put it on my list to further investigate sometime

im new to NATS so this is all a bit overwhelming still :')

logically i'd say; NATS redelivers till acked and NATS dedups till acked, which sounds waterproof to me.

**Comment by ro0NL** [comment]
TLDR; nats doesnt support 'dedup while queued' out of the box, which is our main goal :joy_cat:

**Comment by jnmoyne** [comment]
> TLDR; nats doesnt support 'dedup while queued' out of the box, which is our main goal 😹

It is possible, see my answer above (combination of using a WQQ stream (or removing the messages explicitly) and the 'infinite deduplication window' technique.

**Comment by ro0NL** [comment]
@jnmoyne you're right! it wasnt as intuitive as i imagined, but i've re-read the blog and OK :+1: 

> the more individual subjects you have in a stream, the more the overhead (i.e. memory) taken up on the servers for this indexing

im a bit worried about production numbers, but it sounds worth testing.
